File: 2020/go/03/03.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

type vec struct {
	x int
	y int
}

func answer1(input string, x int, y int) int {
	input = strings.Replace(input, "\n", "", -1)
	pos := vec{0, 0}
	trees := 0
	for pos.y*31+pos.x < len(input)-1 {
		if string(input[pos.y*31+pos.x]) == "#" {
			trees++
		}
		pos.x += x
		pos.y += y
		if pos.x > 30 {
			pos.x -= 31
		}
	}
	return trees
}

func answer2(input string) int {
	a := answer1(input, 1, 1)
	b := answer1(input, 3, 1)
	c := answer1(input, 5, 1)
	d := answer1(input, 7, 1)
	e := answer1(input, 1, 2)
	return a * b * c * d * e
}

func main() {
	input, _ := ioutil.ReadFile("03input.txt")
	fmt.Println(answer1(string(input), 3, 1))
	fmt.Println(answer2(string(input)))
}

File: 2023/lua/03.lua
require "utils"

local input = readFile("03input.txt")
local partNumberFound = {} -- lista de números encontrados
local down = #splitString(input, lineDelimiter)[1] + 1 -- se suma el salto de línea
local up = -down
local adjacents = {-1, up - 1, up, up + 1, 1, down + 1, down, down - 1}

local function readNumberOfIndex(index)
    local left = index
    local right = index
    local number = 0
    while tonumber(input:sub(left - 1, left - 1)) ~= nil do
        left = left - 1
        number = tonumber(input:sub(left, right))
        if number == nil then
            left = left + 1
            break
        end
    end
    while tonumber(input:sub(right + 1, right + 1)) ~= nil do
        right = right + 1
        number = tonumber(input:sub(left, right))
        if number == nil then
            right = right - 1
            break
        end
    end
    return tonumber(input:sub(left, right)), left, right
end

local function getSumOfParts(index, gears)
    local char = ""
    local isNumber = false
    local left = 0
    local right = 0
    local number = 0
    local total = 0
    local gearParts = {}

    for i = 1, #adjacents do
        if adjacents[i] + index > 0 and adjacents[i] + index <= #input then
            char = input:sub(adjacents[i] + index, adjacents[i] + index)
            isNumber = tonumber(char) ~= nil
            if isNumber then
                if partNumberFound[adjacents[i] + index] == nil then
                    number, left, right = readNumberOfIndex(adjacents[i] + index)
                    for j = left, right do
                        partNumberFound[j] = true
                    end
                    if gears then
                        table.insert(gearParts, number)
                    else
                        total = total + number
                    end

                end
            end
        end
    end
    if gears then
        if #gearParts == 2 then
            return gearParts[1] * gearParts[2]
        else
            return 0
        end
    else
        return total
    end
end

local function answer1()
    partNumberFound = {}
    local char = ""
    local total = 0
    for i = 1, #input do
        char = input:sub(i, i)
        if tonumber(char) == nil and char ~= "." and char ~= "\n" then
            total = total + getSumOfParts(i, false)
        end
    end
    return total
end

local function answer2()
    partNumberFound = {}
    local char = ""
    local total = 0
    for i = 1, #input do
        char = input:sub(i, i)
        if char == "*" then
            total = total + getSumOfParts(i, true)
        end
    end
    return total
end

print("Parte 1:", answer1())
print("Parte 2:", answer2())
